allennlp/training/trainer.py

```diff
@@ -32,6 +32,8 @@
 from allennlp.training.tensorboard_writer import TensorboardWriter
 from allennlp.training.trainer_base import TrainerBase
 
+from apex import amp
```

Comment:
I'm ok with just adding this to our list of dependencies, unless it's super heavy.  Though it should be above, in the same block as the `torch` imports.

Also, can you add a test with a not-`None` `opt_level`, showing that this works?

Reply:
Okay, moved apex import to torch block. Put it at the top because its convention for them to be alphabetically sorted (I think?)

[Here](https://github.com/NVIDIA/apex#linux) is the install process for apex. I will leave it you as to whether this is too heavy or not!

Sure, will work on adding a test either tonight or tomorrow.

Reply:
I guess we can add it to our `setup.py`, and if someone wants to install their own optimized version, that's fine too?  The point is we want to not crash when someone just does `pip install allennlp`.

Reply:
Right, that seems perfectly reasonable to me.

Reply:
Also, I need a little time to get my head around how the testing works in AllenNLP. If it is very easy for one of the AllenNLP devs to add a test for when `opt_level is not None` I could also pass the buck so this can get merged sooner.